Readme (#9)
0ph1uch1/pycjson: README.md

# py-cjson

> py-cjson is a python3 json library, powered by algorithms from cJSON library, providing JSON relevant functionality.  It is designed to offer a familiar interface to those who are accustomed to the built-in Python json module, while providing an improved speed and efficiency for JSON encoding and decoding.

Reviewer: ```suggestion
py-cjson is a python3 json library, powered by algorithms from cJSON library, providing JSON relevant functionality. It is designed to offer a similar interface to those who are accustomed to the built-in Python json module, while providing an improved speed for JSON encoding and decoding.
```

# py-cjson

> py-cjson is a Python 3 implementation of the cJSON library, providing JSON parsing and encoding functionality. It aims to have the same interface and parameters as cJSON.

Reviewer: 忘记说了，这里得给 cjson credit(根据 MIT license), 也就是说我们可以说

> py-cjson is a python3 json library, powered by algorithms from cJSON library, providing JSON relevant fuinctionality. It aims to have same interface as Python json library.

## Relevant
To understand more about the cJSON library and its usage, you can refer to the following resources:

> [UltraJSON](https://github.com/ultrajson/ultrajson): Ultra fast JSON decoder and encoder written in C with Python bindings.  

Reviewer: 为什么提到 ujson

[cJSON](https://github.com/DaveGamble/cJSON): Ultralightweight JSON parser in ANSI C.  
## Installation
To install py-cjson, follow these steps:

1. Clone the repository to your local machine.  
2. Navigate to the cloned repository.  
3. Run the command `python -m pip install .` to install the package.

> 

Reviewer: TBD `pip install cjson`

## Usage  
Once py-cjson is installed, you can use it in your Python 3 projects. The interface and parameters provided by py-cjson are designed to be identical to the cJSON library. You can refer to the cJSON documentation for more details on how to use the library.

> Example usage:

Reviewer: 可以给1,2个真实的例子，然后我们的接口参数是设计的和 stdjson 也就是标准 json 库相似

```python
import cjson

# JSON encoding
data = {"key": "value"}
json_str = cjson.encode(data)
print(json_str)

# JSON decoding
decoded_data = cjson.decode(json_str)
print(decoded_data)
```
Please note that the above usage example is a simplified demonstration. Refer to the cJSON documentation for a complete understanding of the library's features and usage patterns.

> ## Available api and parameters

Reviewer: 加上参数名, 参考 stdjson 的文档 <https://docs.python.org/3.11/library/json.html#json.dumps>

[json.org](https://www.json.org/): The official JSON website.  
[Python json](https://docs.python.org/3/library/json.html): The official Python json library documentation.  

## Usage  
Once py-cjson is installed, you can use it in your Python 3 projects. The interface and parameters provided by py-cjson are designed to closely resemble thouse of the JSON library. You can refer to the JSON documentation for more details on how to use the library.

Example usage:
```python
import cjson
# loads, translate json string to python object
cjson.loads(json_str)

# dumps, translate python object to json string
cjson.dumps(data)

# encode same as dumps
json_str = cjson.encode(data)

# decode same as loads
decoded_data = cjson.decode(json_str)

# load from file
with open("data.json", "r") as file:
    data = cjson.load(file)

# dump to file
with open("data.json", "w") as file:
    cjson.dump(data, file)
```

Please note that the above usage example is a simplified demonstration. Refer to the cJSON documentation for a complete understanding of the library's features and usage patterns.

## Available api and parameters
### encode
```python

> cjson.encode(obj, format=None, skipkeys=None, allow_nan=None, separators=None, defaule=None) #Converts arbitrary object recursively into JSON.
> cjson.dumps(obj, format=None, skipkeys=None, allow_nan=None, separators=None, defaule=None) #Converts arbitrary object recursively into JSON.
> cjson.dump(obj, fp, format=None, skipkeys=None, allow_nan=None, separators=None, defaule=None) #Converts arbitrary object recursively into JSON file.

Reviewer: ```suggestion
cjson.encode(obj, *, format=False, skipkeys=False, allow_nan=True, separators=(",",":"), default=None) #Converts arbitrary object recursively into JSON.
cjson.dumps(obj, *, format=False, skipkeys=False, allow_nan=True, separators=(",",":"), default=None) #Converts arbitrary object recursively into JSON.
cjson.dump(obj, fp, *, format=False, skipkeys=False, allow_nan=True, separators=(",",":"), default=None) #Converts arbitrary object recursively into JSON file.
```

# dump optional parameter: default, used to implement custom encoders
def default(obj):
    if isinstance(obj, str):
        return {"key": obj}
    return obj
cjson.dump(data, file, default=default)

# dump optional parameter: allow_nan
data = {"key": float("nan")}
json_str = cjson.encode(data, allow_nan=Fasle) # raises ValueError

# dump optional parameter: separators
json_str = cjson.encode(data, separators=(",", ":"))

# dump optional parameter: skipkeys
data = {"key": float("nan")}
json_str = cjson.encode(data, skipkeys=True) # skips the key

```

### decode
```python

> cjson.decode(s, object_hook=None) #Converts JSON as string to dict object structure.
> cjson.loads(s, object_hook=None) #Converts JSON as string to dict object structure.
> cjson.dump(fp, object_hook=None) #Converts JSON as file to dict object structure.

Reviewer: ```suggestion
cjson.decode(s, *, object_hook=None) #Converts JSON as string to dict object structure.
cjson.loads(s, *, object_hook=None) #Converts JSON as string to dict object structure.
cjson.dump(fp, *, object_hook=None) #Converts JSON as file to dict object structure.
```
